BEP191.md

```diff
@@ -0,0 +1,204 @@
+
+1. Each validator needs to cache the block generation status of nearly 21 block heights, and the historical blocks with a height of more than 21 blocks will be eliminated;
+2. When receiving an `INTURN` block with a height `H`, first verify whether the block timestamp meets expectations, and then calculate the `DelayTime`:
+
+```bash
+DelayTime = receiveTime - parentBlockTime - blockPeriod; 
+//readjustment DelayTime
+if DelayTime < initialBackOffTime: 
+    DelayTime = 0 
+if DelayTime > 2 * (blockPeriod + wiggleTime)
+    DelayTime = 2 * (blockPeriod + wiggleTime)
+```
+
+> The first `INTURN` block of height H is expected to be broadcast to the network at `parentBlockTime + blockPeriod + initialBackOffTime` time.
+>
+> The current blockPeriod is 3s, initialBackOffTime is 1s, and wiggleTime is 1s.
+>
+> Because of network delay, network jitter or burst traffic, there will be a some tolerance range. For example, it is still acceptable to broadcast within 1s. If the validator maliciously delays about 1s to package more transactions, there is no room for tolerance at this time , the validator is likely to be punished.
+>
+> Therefore, if the delay is less than `initialBackOffTime`, there is no need to vote for Slash. At the same time, assuming that BSC will generally not reorg more than 2 blocks, the maximum delay time is `2 * (blockPeriod + wiggleTime)`.
+> 
+> All times with millisecond precision.
+
+3. If the weight `DelayTime > 0`, record that the current supposed validator needs to be punished at height H, and waitting to vote;
+4. When it is the validator's turn to produce a block, collect all the delay broadcast records that need to be punished and submit `DelayTime` to the `SlashIndicator` contract in batches:
+    - Check if records between block `H1` and `H2`, `H1` is last producing block, `H2` is current block;
+    - If the validator `V` has already taken effect at the height of delay broadcast slash or missing block slash, skip it;
+    - If the slash is correct and the cumulative `voteCount > N/3*2`, that means majority found delays, then cumulative `SlashScore`:
+    ```bash
+   AvgDelay = Median(DelayTime)
+   SlashScore += AvgDelay * 100 / (blockPeriod + initialBackOffTime)
+    ```
+    - Check if `V` reach `misdemeanor` or `felony` penalty, run punishment logic;
+
+If it is satisfied that the majority thinks that a certain validator has delayed broadcasting behavior, it is necessary to accumulate `SlashScore`. If the delay time is long enough to continue to produce blocks, the penalty for delayed broadcast is the same as the missing block, and less delay will accumulate fewer points, and it will not be punished in the end.
+
+The following figure is the timing diagram of two nodes voting for validator3 in the case of three validators.
```

Comment:
It is not a deterministic information, may not pass the check here: https://github.com/bnb-chain/bsc/blob/b4773e8b5080f37e1c65c083b543f60c895abb70/consensus/parlia/parlia.go#L1205